examples/ecs/derived_state.rs

```diff
@@ -0,0 +1,209 @@
+//! This example illustrates how to use [`DerivedStates`] for more complex state transitions.
```

Comment:
I think we should try and strengthen the motivation and high-level explanation here.

This is not a concept readers will be familiar with, so we should sell them on it here.